File: tobas_setup_assistant/src/tobas_setup_assistant/setting_widgets/imu.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..setup_assistant import SetupAssistant
    from ..parameter_getters import ParamGetterWidget

from overrides import override
from PyQt5.QtWidgets import QVBoxLayout

from ..common import SENSOR_OFFSET_DESCRIPTION
from ..parameter_getters import ParamGetterWidget_SpinBox, ParamGetterWidget_Vector3d, ParamGetterWidget_DoubleSpinBox
from .base_setting import BaseSettingWidget


class ImuWidget(BaseSettingWidget):
    NAME = "IMU"
    TITLE_TEXT = "Define Inertial Measurement Unit"
    ABST_TEXT = ""  # TODO

    def __init__(self, main: SetupAssistant) -> None:
        super().__init__(main)

        self._param_rows = QVBoxLayout()
        self._rows.addLayout(self._param_rows)

        self.offset = ParamGetterWidget_Vector3d("Offset", SENSOR_OFFSET_DESCRIPTION, suffix=" m")
        self._param_rows.addWidget(self.offset)

        update_rate_description = ""
        self.update_rate = ParamGetterWidget_SpinBox(
            "Update rate", update_rate_description, minimum=1, default=400, suffix=" Hz"
        )
        self._param_rows.addWidget(self.update_rate)

        gyro_noise_density_description = ""
        self.gyro_noise_density = ParamGetterWidget_DoubleSpinBox(
            "Gyroscope noise density (two-sided spectrum)",
            gyro_noise_density_description,
            decimals=9,
            minimum=0.0,
            default=0.005,
            suffix=" rad/s/sqrt(Hz)",
        )
        self._param_rows.addWidget(self.gyro_noise_density)

        gyro_random_walk_description = ""
        self.gyro_random_walk = ParamGetterWidget_DoubleSpinBox(
            "Gyroscope bias random walk",
            gyro_random_walk_description,
            decimals=9,
            minimum=0.0,
            default=1e-4,
            suffix=" rad/s^2/sqrt(Hz)",
        )
        self._param_rows.addWidget(self.gyro_random_walk)

        gyro_bias_corr_time_description = ""
        self.gyro_bias_corr_time = ParamGetterWidget_SpinBox(
            "Gyroscope bias correlation time constant",
            gyro_bias_corr_time_description,
            minimum=0,
            default=1000,
            suffix=" s",
        )
        self._param_rows.addWidget(self.gyro_bias_corr_time)

        gyro_turn_on_bias_sigma_description = ""
        self.gyro_turn_on_bias_sigma = ParamGetterWidget_DoubleSpinBox(
            "Gyroscope turn on bias standard deviation",
            gyro_turn_on_bias_sigma_description,
            decimals=9,
            minimum=0.0,
            default=0.05,
            suffix=" rad/s",
        )
        self._param_rows.addWidget(self.gyro_turn_on_bias_sigma)

        gyro_lpf_cutoff_freq_description = ""
        self.gyro_lpf_cutoff_freq = ParamGetterWidget_SpinBox(
            "Gyroscope LPF cutoff frequency",
            gyro_lpf_cutoff_freq_description,
            minimum=1,
            maximum=400,
            default=20,
            suffix=" Hz",
        )
        self._param_rows.addWidget(self.gyro_lpf_cutoff_freq)

        acc_noise_density_description = ""
        self.acc_noise_density = ParamGetterWidget_DoubleSpinBox(
            "Accelerometer noise density (two-sided spectrum)",
            acc_noise_density_description,
            decimals=9,
            minimum=0.0,
            default=0.05,
            suffix=" m/s^2/sqrt(Hz)",
        )
        self._param_rows.addWidget(self.acc_noise_density)

        acc_random_walk_description = ""
        self.acc_random_walk = ParamGetterWidget_DoubleSpinBox(
            "Accelerometer bias random walk",
            acc_random_walk_description,
            decimals=9,
            minimum=0.0,
            default=0.01,
            suffix=" m/s^3/sqrt(Hz)",
        )
        self._param_rows.addWidget(self.acc_random_walk)

        acc_bias_corr_time_description = ""
        self.acc_bias_corr_time = ParamGetterWidget_SpinBox(
            "Accelerometer bias correlation time constant",
            acc_bias_corr_time_description,
            minimum=0,
            default=300,
            suffix=" s",
        )
        self._param_rows.addWidget(self.acc_bias_corr_time)

        acc_turn_on_bias_sigma_description = ""
        self.acc_turn_on_bias_sigma = ParamGetterWidget_DoubleSpinBox(
            "Accelerometer turn on bias standard deviation",
            acc_turn_on_bias_sigma_description,
            decimals=9,
            minimum=0.0,
            default=0.2,
            suffix=" m/s^2",
        )
        self._param_rows.addWidget(self.acc_turn_on_bias_sigma)

        acc_lpf_cutoff_freq_description = ""
        self.acc_lpf_cutoff_freq = ParamGetterWidget_SpinBox(
            "Accelerometer LPF cutoff frequency",
            acc_lpf_cutoff_freq_description,
            minimum=1,
            maximum=400,
            default=20,
            suffix=" Hz",
        )
        self._param_rows.addWidget(self.acc_lpf_cutoff_freq)

        mag_gauss_noise_description = ""
        self.mag_gauss_noise = ParamGetterWidget_SpinBox(
            "Magnetometer standard deviation of additive white gaussian noise",
            mag_gauss_noise_description,
            minimum=0,
            default=80,
            suffix=" nT",
        )
        self._param_rows.addWidget(self.mag_gauss_noise)

        mag_uniform_noise_description = ""
        self.mag_uniform_noise = ParamGetterWidget_SpinBox(
            "Magnetometer symmetric bounds of uniform noise for initial sensor bias",
            mag_uniform_noise_description,
            minimum=0,
            default=400,
            suffix=" nT",
        )
        self._param_rows.addWidget(self.mag_uniform_noise)

        self._rows.addStretch()

    @override
    def update_internal_data_structures(self) -> None:
        pass

    @override
    def is_valid(self) -> bool:
        return True

    @override
    def dump_settings(self) -> dict:
        res = dict()
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            res[param.name()] = param.get()
        return res

    @override
    def load_settings(self, data: dict) -> None:
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            param.set(data[param.name()])

    def equipped(self) -> bool:
        return True

File: tobas_gui/tobas_setup_assistant/src/setting_widgets/imu.cpp
#include <tobas_std_tools/universal_constants.hpp>
#include <tobas_yaml_tools/convert/eigen.hpp>

#include "tobas_setup_assistant/setting_tabs/imu.hpp"

namespace gui
{
namespace sa
{
IMUWidget::IMUWidget()
{
  offset_ = new ParamGetterWidget_Vector3d("Offset", kSensorOffsetDescription);
  offset_->setSuffix(" m");
  addWidget(offset_);

  update_rate_ = new ParamGetterWidget_SpinBox("Update Rate", "");  // TODO
  update_rate_->setMinimum(1);
  update_rate_->setValue(400);
  update_rate_->setSuffix(" Hz");
  addWidget(update_rate_);

  gyro_noise_density_ = new ParamGetterWidget_SpinBox("Gyroscope Noise Density", "");  // TODO
  gyro_noise_density_->setMinimum(0);
  gyro_noise_density_->setValue(11);  // ISM330DLC
  gyro_noise_density_->setSuffix(" mdps/√Hz");
  addWidget(gyro_noise_density_);

  gyro_offset_norm_ = new ParamGetterWidget_SpinBox("Gyroscope Initial Offset Norm", "");  // TODO
  gyro_offset_norm_->setMinimum(0);
  gyro_offset_norm_->setValue(2);  // ISM330DLC
  gyro_offset_norm_->setSuffix(" dps");
  addWidget(gyro_offset_norm_);

  gyro_random_walk_ = new ParamGetterWidget_DoubleSpinBox("Gyroscope Bias Random Walk", "");  // TODO
  gyro_random_walk_->setDecimals(6);
  gyro_random_walk_->setMinimum(0.);
  gyro_random_walk_->setValue(0.);  // TODO
  gyro_random_walk_->setSuffix(" rad/s^2/√Hz");
  addWidget(gyro_random_walk_);

  gyro_bias_corr_time_ = new ParamGetterWidget_SpinBox("Gyroscope Bias Correlation Time Constant", "");  // TODO
  gyro_bias_corr_time_->setMinimum(0);
  gyro_bias_corr_time_->setValue(1000);  // TODO
  gyro_bias_corr_time_->setSuffix(" s");
  addWidget(gyro_bias_corr_time_);

  acc_noise_density_ = new ParamGetterWidget_SpinBox("Accelerometer Noise Density", "");  // TODO
  acc_noise_density_->setMinimum(0);
  acc_noise_density_->setValue(170);  // ISM330DLC
  acc_noise_density_->setSuffix(" ug/√Hz");
  addWidget(acc_noise_density_);

  acc_offset_norm_ = new ParamGetterWidget_SpinBox("Accelerometer Initial Offset Norm", "");  // TODO
  acc_offset_norm_->setMinimum(0.);
  acc_offset_norm_->setValue(40);  // ISM330DLC
  acc_offset_norm_->setSuffix(" mg");
  addWidget(acc_offset_norm_);

  acc_random_walk_ = new ParamGetterWidget_DoubleSpinBox("Accelerometer Bias Random Walk", "");  // TODO
  acc_random_walk_->setDecimals(6);
  acc_random_walk_->setMinimum(0.);
  acc_random_walk_->setValue(0.);  // TODO
  acc_random_walk_->setSuffix(" m/s^3/√Hz");
  addWidget(acc_random_walk_);

  acc_bias_corr_time_ = new ParamGetterWidget_SpinBox("Accelerometer Bias Correlation Time Constant", "");  // TODO
  acc_bias_corr_time_->setMinimum(0);
  acc_bias_corr_time_->setValue(300);  // TODO
  acc_bias_corr_time_->setSuffix(" s");
  addWidget(acc_bias_corr_time_);

  addStretch();
}

const char* IMUWidget::name() const
{
  return "IMU";
}

const char* IMUWidget::title() const
{
  return "Define 6-Axis IMU";
}

const char* IMUWidget::description() const
{
  return "";  // TODO
}

void IMUWidget::onOpened()
{
  return;
}

void IMUWidget::updateInternalDataStructures()
{
  return;
}

bool IMUWidget::isValid()
{
  return true;
}

YAML::Node IMUWidget::dump()
{
  YAML::Node node(YAML::NodeType::Map);

  node[offset_->name()] = offset_->getValue();
  node[update_rate_->name()] = update_rate_->getValue();
  node[gyro_noise_density_->name()] = gyro_noise_density_->getValue();
  node[gyro_offset_norm_->name()] = gyro_offset_norm_->getValue();
  node[gyro_random_walk_->name()] = gyro_random_walk_->getValue();
  node[gyro_bias_corr_time_->name()] = gyro_bias_corr_time_->getValue();
  node[acc_noise_density_->name()] = acc_noise_density_->getValue();
  node[acc_offset_norm_->name()] = acc_offset_norm_->getValue();
  node[acc_random_walk_->name()] = acc_random_walk_->getValue();
  node[acc_bias_corr_time_->name()] = acc_bias_corr_time_->getValue();

  return node;
}

void IMUWidget::load(const YAML::Node& node)
{
  offset_->setValue(node[offset_->name()].as<Eigen::Vector3d>());
  update_rate_->setValue(node[update_rate_->name()].as<int>());
  gyro_noise_density_->setValue(node[gyro_noise_density_->name()].as<double>());
  gyro_offset_norm_->setValue(node[gyro_offset_norm_->name()].as<double>());
  gyro_random_walk_->setValue(node[gyro_random_walk_->name()].as<double>());
  gyro_bias_corr_time_->setValue(node[gyro_bias_corr_time_->name()].as<int>());
  acc_noise_density_->setValue(node[acc_noise_density_->name()].as<double>());
  acc_offset_norm_->setValue(node[acc_offset_norm_->name()].as<double>());
  acc_random_walk_->setValue(node[acc_random_walk_->name()].as<double>());
  acc_bias_corr_time_->setValue(node[acc_bias_corr_time_->name()].as<int>());
}

Eigen::Vector3d IMUWidget::offset() const
{
  return offset_->getValue();
}

int IMUWidget::updateRate() const
{
  return update_rate_->getValue();
}

double IMUWidget::gyroNoiseDensity() const
{
  return (tobas_std::kDeg2Rad * 1e-3) * gyro_noise_density_->getValue();
}

double IMUWidget::gyroOffsetNorm() const
{
  return tobas_std::kDeg2Rad * gyro_offset_norm_->getValue();
}

double IMUWidget::gyroRandomWalk() const
{
  return gyro_random_walk_->getValue();
}

int IMUWidget::gyroBiasCorrTime() const
{
  return gyro_bias_corr_time_->getValue();
}

double IMUWidget::accNoiseDensity() const
{
  return (tobas_std::kGravity * 1e-6) * acc_noise_density_->getValue();
}

double IMUWidget::accOffsetNorm() const
{
  return (tobas_std::kGravity * 1e-3) * acc_offset_norm_->getValue();
}

double IMUWidget::accRandomWalk() const
{
  return acc_random_walk_->getValue();
}

int IMUWidget::accBiasCorrTime() const
{
  return acc_bias_corr_time_->getValue();
}
}  // namespace sa
}  // namespace gui
